qiskit_experiments/framework/backend_timing.py

```diff
@@ -0,0 +1,363 @@
+        the time until the next valid pulse, assuming the ``Delay`` instruction
+        begins on a sample that is also valid for a pulse to begin on.
+
+        The pulse timing constraints of the backend are considered in order to
+        give the number of samples closest to ``time`` for the start of a pulse
+        in a subsequent instruction to be valid. The least common multiple of
+        the pulse and acquire alignment values is used in order to ensure that
+        either type of pulse will be aligned.
+
+        If :meth:`.BackendTiming.delay_unit` is ``s``, ``time`` is
+        returned directly. Typically, this is the case for a simulator where
+        converting to sample number is not needed.
+
+        Args:
+            time: The nominal delay time to convert in seconds
+
+        Returns:
+            The delay duration in samples if :meth:`.BackendTiming.delay_unit`
+            is ``dt``. Otherwise return ``time``.
+        """
+        if self.delay_unit == "s":
+            return time
+
+        return self.schedule_delay(time)
+
+    def schedule_delay(self, time: float) -> int:
+        """Closest valid delay in samples to ``time``
+
+        The pulse timing constraints of the backend are considered in order to
+        give a number of samples closest to ``time`` consistent with the
+        alignment constraints, so that the start of a pulse in a subsequent
+        instruction is valid. The least common multiple of the pulse and
+        acquire alignment values is used in order to ensure that either type of
+        pulse will be aligned.
+
+        Args:
+            time: The nominal delay time to convert in seconds
+
+        Returns:
+            The delay duration in samples to pass to a ``Delay`` instruction in
+            Qiskit pulse schedule
+
+        Raises:
+            QiskitError: The backend does not include a dt value.
+        """
+        if self.dt is None:
+            raise QiskitError("Backend has no dt value.")
+
+        granularity = lcm(self._pulse_alignment, self._acquire_alignment)
+
+        samples = int(round(time / self.dt / granularity) * granularity)
+
+        return samples
+
+    def pulse_samples(self, time: float) -> int:
+        """The number of samples giving the valid pulse duration closest to ``time``
+
+        The multiple of the pulse granularity giving the time closest to
+        ``time`` is used. The returned value is always at least the backend's
+        ``min_length``.
+
+        Args:
+            time: Pulse duration in seconds
+
+        Returns:
+            The number of samples corresponding to ``time``
+
+        Raises:
+            QiskitError: The backend does not include a dt value.
+        """
+        if self.dt is None:
+            raise QiskitError("Backend has no dt value.")
+
+        return self.round_pulse_samples(time / self.dt)
+
+    def round_pulse_samples(self, samples: Union[float, int]) -> int:
```

Comment:
Since `pulse_samples` also rounds the value, the difference of two methods is not clear without reading the code. To me `pulse_samples(self, *, sample=None, time=None)` is more readable (enforcing a coder to write argument name is also good for reading QE code, because usually unit of scanned times in the experimental option is not clear. Maybe we can do the same interface for circuit delay.

Reply:
Okay, I was thinking "round" implied adjusting the number without changing units but the other interface seems fine as well, so I will switch to that.

Reply:
Methods refactored in [d531cc1](https://github.com/Qiskit/qiskit-experiments/pull/860/commits/d531cc1ed4f85136153096f090c4e041d1f8ec5b).